File: android/app/src/main/java/io/github/shiot0/shou/ShouStore.kt
package io.github.shiot0.shou

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import org.json.JSONArray
import org.json.JSONObject

/** One saved Shou server, mirrored from the web remote's localStorage set. */
data class Remote(
    val id: String,
    val name: String,
    val key: String,
    val host: String,
    val hostname: String,
    val port: String,
    val mac: String,
) {
    /** Best address to reach this server on right now. */
    fun bestHost(): String = host.ifBlank { hostname }

    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id); put("name", name); put("key", key)
        put("host", host); put("hostname", hostname); put("port", port); put("mac", mac)
    }

    companion object {
        fun fromJson(o: JSONObject) = Remote(
            id = o.optString("id"),
            name = o.optString("name"),
            key = o.optString("key"),
            host = o.optString("host"),
            hostname = o.optString("hostname"),
            port = o.optString("port", "4100").ifBlank { "4100" },
            mac = o.optString("mac"),
        )
    }
}

/** A compact mirror of what the kiosk is doing, pushed from the web remote each tick. */
data class Playback(
    val active: Boolean,   // something is playing (mpv is up)
    val playing: Boolean,  // true = playing, false = paused
    val title: String,
    val subtitle: String,
    val cover: String,
    val positionMs: Long,
    val durationMs: Long,
)

/**
 * The native source of truth shared by every background entry point (media session,
 * widget, Quick Settings tile, Wake-on-LAN, shortcuts). Two stores:
 *
 *  * a hardware-encrypted store for anything secret — the saved remotes (which carry
 *    their REMOTE_TOKEN keys) and the single-server token from Settings;
 *  * a plain store for the non-secret bits the widget/tile need to render quickly
 *    (active host/port/name + the last playback snapshot).
 */
object ShouStore {

    private const val SECURE = "shou_secure"
    private const val PLAIN = "shou"          // shared with SettingsActivity / MainActivity

    @Volatile private var securePrefs: SharedPreferences? = null
    @Volatile private var plainPrefs: SharedPreferences? = null

    fun init(ctx: Context) {
        val app = ctx.applicationContext
        if (plainPrefs == null) plainPrefs = app.getSharedPreferences(PLAIN, Context.MODE_PRIVATE)
        if (securePrefs == null) securePrefs = buildSecure(app)
    }

    private fun secure(ctx: Context): SharedPreferences {
        securePrefs?.let { return it }
        return buildSecure(ctx.applicationContext).also { securePrefs = it }
    }

    private fun plain(ctx: Context): SharedPreferences {
        plainPrefs?.let { return it }
        return ctx.applicationContext.getSharedPreferences(PLAIN, Context.MODE_PRIVATE)
            .also { plainPrefs = it }
    }

    /** Encrypted prefs, or a graceful fall back to plain prefs if the keystore misbehaves
     *  (some OEM builds throw on EncryptedSharedPreferences) — the app must still work. */
    private fun buildSecure(app: Context): SharedPreferences = try {
        val key = MasterKey.Builder(app)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            app, SECURE, key,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    } catch (e: Throwable) {
        app.getSharedPreferences(SECURE + "_fallback", Context.MODE_PRIVATE)
    }

    // --- Single-server settings (Settings screen) --------------------------- //
    // host / port / https / allowBadCerts stay in plain prefs (not secret); only the
    // token is sensitive, so it lives in the encrypted store.

    fun host(ctx: Context): String = plain(ctx).getString("host", "")!!.trim()
    fun port(ctx: Context): String = plain(ctx).getString("port", "4100")!!.trim().ifEmpty { "4100" }
    fun https(ctx: Context): Boolean = plain(ctx).getBoolean("https", false)
    fun allowBadCerts(ctx: Context): Boolean = plain(ctx).getBoolean("allowBadCerts", false)

    fun token(ctx: Context): String {
        // Read the encrypted token, migrating a legacy plaintext token on first run.
        val s = secure(ctx)
        s.getString("token", null)?.let { return it }
        val legacy = plain(ctx).getString("token", "")!!.trim()
        if (legacy.isNotEmpty()) {
            s.edit().putString("token", legacy).apply()
            plain(ctx).edit().remove("token").apply()
        }
        return legacy
    }

    fun saveSettings(
        ctx: Context, host: String, port: String, token: String,
        https: Boolean, allowBadCerts: Boolean,
    ) {
        plain(ctx).edit()
            .putString("host", host.trim())
            .putString("port", port.trim().ifEmpty { "4100" })
            .putBoolean("https", https)
            .putBoolean("allowBadCerts", allowBadCerts)
            .remove("token")  // never keep it in plaintext
            .apply()
        secure(ctx).edit().putString("token", token.trim()).apply()
    }

    // --- Saved remotes (synced from the web remote's localStorage set) ------- //

    fun setRemotes(ctx: Context, json: String) {
        val cleaned = try {
            val arr = JSONArray(json)
            val out = JSONArray()
            for (i in 0 until arr.length()) {
                val o = arr.optJSONObject(i) ?: continue
                if (o.optString("key").isBlank()) continue
                out.put(Remote.fromJson(o).toJson())
            }
            out.toString()
        } catch (e: Exception) {
            return
        }
        secure(ctx).edit().putString("remotes", cleaned).apply()
    }

    fun remotes(ctx: Context): List<Remote> {
        val raw = secure(ctx).getString("remotes", null) ?: return emptyList()
        return try {
            val arr = JSONArray(raw)
            (0 until arr.length()).mapNotNull { arr.optJSONObject(it)?.let(Remote::fromJson) }
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun remoteByToken(ctx: Context, token: String): Remote? =
        remotes(ctx).firstOrNull { it.key == token }

    // --- The active endpoint (what background features talk to) -------------- //
    // Mirrored from the web on every load, so the media session / widget / WOL keep
    // working even after the user switched remotes inside the WebView.

    fun setActive(ctx: Context, token: String, host: String, port: String, name: String) {
        secure(ctx).edit().putString("active_token", token).apply()
        plain(ctx).edit()
            .putString("active_host", host)
            .putString("active_port", port.ifBlank { "4100" })
            .putString("active_name", name)
            .apply()
    }

    fun activeToken(ctx: Context): String =
        secure(ctx).getString("active_token", null)?.takeIf { it.isNotBlank() } ?: token(ctx)

    fun activeHost(ctx: Context): String =
        plain(ctx).getString("active_host", null)?.takeIf { it.isNotBlank() } ?: host(ctx)

    fun activePort(ctx: Context): String =
        plain(ctx).getString("active_port", null)?.takeIf { it.isNotBlank() } ?: port(ctx)

    fun activeName(ctx: Context): String =
        plain(ctx).getString("active_name", null)?.takeIf { it.isNotBlank() }
            ?: activeHost(ctx).ifBlank { "Shou" }

    /** Base http(s) URL for the active server, or null if nothing is configured yet. */
    fun activeBaseUrl(ctx: Context): String? {
        val host = activeHost(ctx)
        if (host.isBlank()) return null
        val scheme = if (https(ctx)) "https" else "http"
        return "$scheme://$host:${activePort(ctx)}"
    }

    /** MAC for Wake-on-LAN of the active server (from its saved remote), or "". */
    fun activeMac(ctx: Context): String =
        remoteByToken(ctx, activeToken(ctx))?.mac?.trim().orEmpty()

    // --- Playback snapshot (plain — refreshed ~1/sec, read by widget & tile) -- //

    fun setPlayback(ctx: Context, p: Playback?) {
        val e = plain(ctx).edit()
        if (p == null || !p.active) {
            e.putBoolean("pb_active", false)
        } else {
            e.putBoolean("pb_active", true)
                .putBoolean("pb_playing", p.playing)
                .putString("pb_title", p.title)
                .putString("pb_subtitle", p.subtitle)
                .putString("pb_cover", p.cover)
                .putLong("pb_pos", p.positionMs)
                .putLong("pb_dur", p.durationMs)
        }
        e.apply()
    }

    fun playback(ctx: Context): Playback? {
        val p = plain(ctx)
        if (!p.getBoolean("pb_active", false)) return null
        return Playback(
            active = true,
            playing = p.getBoolean("pb_playing", true),
            title = p.getString("pb_title", "") ?: "",
            subtitle = p.getString("pb_subtitle", "") ?: "",
            cover = p.getString("pb_cover", "") ?: "",
            positionMs = p.getLong("pb_pos", 0L),
            durationMs = p.getLong("pb_dur", 0L),
        )
    }
}

File: ios/Shou/Core/ShouStore.swift
//  ShouStore.swift
//  The native source of truth shared by every entry point (WebView bridge, media
//  session, widget, AppIntents, Wake-on-LAN, shortcuts, background airing). Ported
//  from ShouStore.kt. Two backing stores:
//
//   * the Keychain (shared via the app's keychain-access-group) for anything secret
//     — the saved remotes (which carry REMOTE_TOKEN keys), the single-server token,
//     and the active token; reachable from the widget process too.
//   * an App Group UserDefaults for the non-secret bits the widget/intents read to
//     render fast — active host/port/name + the last playback snapshot.

import Foundation
import Security

enum ShouStore {

    static let appGroup = "group.io.github.shiot0.shou"
    private static let keychainService = "io.github.shiot0.shou"

    private static var defaults: UserDefaults {
        UserDefaults(suiteName: appGroup) ?? .standard
    }

    // MARK: - Single-server settings (Settings screen) ----------------------- //

    static func host() -> String { defaults.string(forKey: "host")?.trimmed ?? "" }
    static func port() -> String {
        let p = defaults.string(forKey: "port")?.trimmed ?? "4100"
        return p.isEmpty ? "4100" : p
    }
    static func https() -> Bool { defaults.bool(forKey: "https") }
    static func allowBadCerts() -> Bool { defaults.bool(forKey: "allowBadCerts") }
    /// Present for Settings parity with Android; inert on iOS (no volume-key API).
    static func volumeKeys() -> Bool { defaults.object(forKey: "volumeKeys") as? Bool ?? false }

    static func token() -> String { Keychain.get("token") ?? "" }

    static func saveSettings(host: String, port: String, token: String,
                             https: Bool, allowBadCerts: Bool, volumeKeys: Bool = false) {
        let d = defaults
        d.set(host.trimmed, forKey: "host")
        let p = port.trimmed
        d.set(p.isEmpty ? "4100" : p, forKey: "port")
        d.set(https, forKey: "https")
        d.set(allowBadCerts, forKey: "allowBadCerts")
        d.set(volumeKeys, forKey: "volumeKeys")
        Keychain.set("token", token.trimmed)
    }

    // MARK: - Saved remotes (synced from the web remote's localStorage set) --- //

    static func setRemotes(json: String) {
        guard let data = json.data(using: .utf8),
              let list = try? JSONDecoder().decode([Remote].self, from: data) else { return }
        let cleaned = list.filter { !$0.key.isEmpty }
        guard let out = try? JSONEncoder().encode(cleaned),
              let str = String(data: out, encoding: .utf8) else { return }
        Keychain.set("remotes", str)
    }

    static func remotes() -> [Remote] {
        guard let raw = Keychain.get("remotes"), let data = raw.data(using: .utf8),
              let list = try? JSONDecoder().decode([Remote].self, from: data) else { return [] }
        return list
    }

    static func remote(byToken token: String) -> Remote? {
        remotes().first { $0.key == token }
    }

    // MARK: - Active endpoint (what background features talk to) -------------- //

    static func setActive(token: String, host: String, port: String, name: String) {
        Keychain.set("active_token", token)
        let d = defaults
        d.set(host, forKey: "active_host")
        d.set(port.isEmpty ? "4100" : port, forKey: "active_port")
        d.set(name, forKey: "active_name")
    }

    static func activeToken() -> String {
        let t = Keychain.get("active_token") ?? ""
        return t.isEmpty ? token() : t
    }
    static func activeHost() -> String {
        let h = defaults.string(forKey: "active_host") ?? ""
        return h.isEmpty ? host() : h
    }
    static func activePort() -> String {
        let p = defaults.string(forKey: "active_port") ?? ""
        return p.isEmpty ? port() : p
    }
    static func activeName() -> String {
        let n = defaults.string(forKey: "active_name") ?? ""
        if !n.isEmpty { return n }
        let h = activeHost()
        return h.isEmpty ? "Shou" : h
    }

    /// Base http(s) URL for the active server, or nil if nothing is configured yet.
    static func activeBaseURL() -> String? {
        let h = activeHost()
        guard !h.isEmpty else { return nil }
        let scheme = https() ? "https" : "http"
        return "\(scheme)://\(h):\(activePort())"
    }

    /// MAC for Wake-on-LAN of the active server (from its saved remote), or "".
    static func activeMac() -> String {
        remote(byToken: activeToken())?.mac.trimmed ?? ""
    }

    // MARK: - Playback snapshot (App Group — read by widget & intents) -------- //

    static func setPlayback(_ p: Playback?) {
        let d = defaults
        guard let p = p, p.active else {
            d.set(false, forKey: "pb_active"); return
        }
        d.set(true, forKey: "pb_active")
        d.set(p.playing, forKey: "pb_playing")
        d.set(p.title, forKey: "pb_title")
        d.set(p.subtitle, forKey: "pb_subtitle")
        d.set(p.cover, forKey: "pb_cover")
        d.set(p.position, forKey: "pb_pos")
        d.set(p.duration, forKey: "pb_dur")
        d.set(Date().timeIntervalSince1970, forKey: "pb_stamp")
    }

    static func playback() -> Playback? {
        let d = defaults
        guard d.bool(forKey: "pb_active") else { return nil }
        return Playback(
            active: true,
            playing: d.bool(forKey: "pb_playing"),
            title: d.string(forKey: "pb_title") ?? "",
            subtitle: d.string(forKey: "pb_subtitle") ?? "",
            cover: d.string(forKey: "pb_cover") ?? "",
            position: d.double(forKey: "pb_pos"),
            duration: d.double(forKey: "pb_dur")
        )
    }

    /// Wall-clock seconds since the snapshot was written (used to extrapolate the
    /// widget progress bar between timeline refreshes).
    static func playbackAge() -> TimeInterval {
        let stamp = defaults.double(forKey: "pb_stamp")
        return stamp > 0 ? Date().timeIntervalSince1970 - stamp : 0
    }
}

private extension String {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}

// MARK: - Minimal shared-Keychain wrapper -------------------------------------- //
// No explicit kSecAttrAccessGroup: both the app and the widget declare the same
// single keychain-access-group, so items default into it and are shared.

enum Keychain {
    private static let service = "io.github.shiot0.shou"

    static func get(_ account: String) -> String? {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var out: AnyObject?
        guard SecItemCopyMatching(q as CFDictionary, &out) == errSecSuccess,
              let data = out as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func set(_ account: String, _ value: String) {
        let data = Data(value.utf8)
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        if SecItemCopyMatching(base as CFDictionary, nil) == errSecSuccess {
            SecItemUpdate(base as CFDictionary,
                          [kSecValueData as String: data] as CFDictionary)
        } else {
            var add = base
            add[kSecValueData as String] = data
            add[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
            SecItemAdd(add as CFDictionary, nil)
        }
    }

    static func delete(_ account: String) {
        let q: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(q as CFDictionary)
    }
}
